### PR TITLE
Account management

### PR DESCRIPTION
**Pull request by davidiw**
This PR will eventually do a lot of things:

- [x] Cleanup the API for create accounts
- [x] Only allow for derived address <> auth keys at account creation
- [x] Allow for account creation within a module with globally unique account addresses
- [ ] Capability for `move_to`
- [x] Capability to create a signer (only at the time of account creation)

Not going to do `move_to`  due to limitations in Move, that require `move_to` is used in the same module where the resource is defined. Will likely need a native function to do this. That is outside the scope of this PR.

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 1]
```diff
@@ -343,18 +343,70 @@ module AptosFramework::Account {
...
+    /// Basic account creation methods.
     ///////////////////////////////////////////////////////////////////////////
 
     public(script) fun create_account(auth_key: address) {
-        let (signer, _) = create_account_internal(auth_key);
+        let signer = create_account_internal(auth_key);
         Coin::register<TestCoin>(&signer);
     }
 
+    public fun create_module_account(
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/Account.move:354]: should this be called create_shared_account?  module_account makes me think of the program account in ETH.

Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Account.move:354]: this is a weird function without doc, I was expecting each account owner can extract a signer capability itself and rotate private key to 0x0 or something

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 2]
```diff
@@ -343,18 +343,70 @@ module AptosFramework::Account {
...
+        source: &signer,
+    ): (signer, SignerCapability) {
+        let guid = GUID::create(source);
+        let bytes = BCS::to_bytes(&guid);
+        Vector::append(&mut bytes, BCS::to_bytes(&Timestamp::now_microseconds()));
+        let addr = create_address(Hash::sha3_256(bytes));
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Account.move:360]: Nice! This is the same way contract addresses are created in EVM - hashes of the deployer accounts. Can we potentially use the account's nonce here if that's available? Does GUID::create(source) guarantee uniqueness here among multiple modules belonging to the same account?

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 3]
```diff
@@ -343,18 +343,70 @@ module AptosFramework::Account {
...
     ///////////////////////////////////////////////////////////////////////////
 
     public(script) fun create_account(auth_key: address) {
-        let (signer, _) = create_account_internal(auth_key);
+        let signer = create_account_internal(auth_key);
         Coin::register<TestCoin>(&signer);
     }
 
+    public fun create_module_account(
+        source: &signer,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Account.move:355]: Can we consider calling this function as part of deploying a module - basically when someone deploy a module, we create a brand new separate account for that module and then hand the SignerCapability to the account deploying it by default? This is definitely a lot more work but can make things a bit easier.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 1 above)
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Account.move:354]: This function is misleading - create_module_account can actually be called to create accounts that do not need to host modules. They can be used as normal accounts. I think we should consider restricting the scope of this to just modules. Being able to freely create accounts that users don't have keys for that are not for modules only can have wider security and trust implications that we might not be able to fully anticipate.

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 4]
```diff
@@ -34,6 +35,8 @@ module AptosFramework::Account {
         currency_code_required: bool,
     }
 
+    struct SignerCapability has copy, drop, store { account: address }
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Account.move:38]: +1. This can be dangerous. I think if we keep the scope to modules only, we should consider renaming this to ModuleOwnerCapability.

[the remark this answers is not in the document]
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:38]: copy definitely
I'm fine with ModuleOwner but it is a SignerCapability :P

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 1 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:354]: That's exactly the goal :P

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:354]: No, let's not do that. Let's make this a part of a complex app registration. Might be a good chance for @areshand initialize function ;).

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:354]: I don't disagree, but the problem is that we'd have to make some substantial changes to the adapter or the account resource to limit that and it isn't entirely clear to me what the implications would be here. What are the problems we want to solve?

Given that we need to start stimulating the community to build, I think we should keep this in mind, but move forward.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 3 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:355]: You mean as part of the initialize function? Maybe we could move down that path. Honestly though, I'm not that concerned. These accounts are truly randomly distributed in the system. There's no more additional cost than adding a new entry to a table.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 2 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:360]: we could eliminate the timestamp in microseconds if we want that. I think we don't need that as I can imagine the flow would be:

```
(signer, cap) = create_module_account()
dao = Dao { signer_cap: cap, .... };
move_to(&signer, dao)
```

**Review by davidiw — COMMENTED** [review]
My proposal is:
* Remove copy
* Add a comment with an example of how this might be used
* Maybe call it create_resource_account instead of create_module_account

land quickly and share with the community so they can start hacking on DAOs and other fun ideas.

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 5]
```diff
@@ -34,6 +35,8 @@ module AptosFramework::Account {
         currency_code_required: bool,
     }
 
+    struct SignerCapability has drop, store { account: address }
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Account.move:38]: Is SignerCap droppable so people can just drop the cap as a way to remove ownership in their script/module code? An alternative here is not to make it droppable and provide a function to destroy SignerCap in Account. I actually prefer the latter more as it makes it very explicit that someone is destroying the SignerCap.

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 6]
```diff
@@ -343,18 +343,95 @@ module AptosFramework::Account {
...
+    /// Basic account creation methods.
     ///////////////////////////////////////////////////////////////////////////
 
     public(script) fun create_account(auth_key: address) {
-        let (signer, _) = create_account_internal(auth_key);
+        let signer = create_account_internal(auth_key);
         Coin::register<TestCoin>(&signer);
     }
 
+    /// A resource account is used to manage resources independent of an account managed by a user.
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/Account.move:354]: Speaking of resources vs modules, can we deploy modules with this account + SignerCap currently? My guess is no right since module bundle upload is a separate path currently.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 3 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Account.move:380]: The reason I think we should explore this use case for modules more, as I also mentioned in earlier comments, is to allow modules to exist separately from an account with private keys. This allows building decentralized systems more easily.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 2 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Account.move:385]: You didn't really answer any of my questions :) but upon looking at GUID, it also uses some concept of counter already.

File: aptos-move/framework/src/natives/account.rs [hunk 7]
```diff
@@ -16,29 +16,36 @@ use move_deps::{
...
+            cost,
+            super::status::NFE_UNABLE_TO_PARSE_ADDRESS,
+        ))
+    }
 }
 
-/// NOTE: this function will be deprecated after the Aptos v3 release, but must
-/// remain for replaying old transactions
-pub fn native_destroy_signer(
+pub fn native_create_signer(
```
Comment (movekevin) [inline aptos-move/framework/src/natives/account.rs:40]: Code diffing in Github is so dumb. This native_create_signer looks exactly the same as before (which was earlier in the file) but it's being shown as a diff.

File: aptos-move/framework/src/natives/account.rs [hunk 8]
```diff
@@ -16,29 +16,36 @@ use move_deps::{
...
     ty_args: Vec<Type>,
     mut arguments: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
     debug_assert!(arguments.len() == 1);
 
-    let address = pop_arg!(arguments, AccountAddress);
-    let cost = native_gas(context.cost_table(), NativeCostIndex::CREATE_SIGNER, 0);
-    Ok(NativeResult::ok(cost, smallvec![Value::signer(address)]))
+    let cost = GasCost::new(super::cost::APTOS_CREATE_ADDRESS, 1).total();
```
Comment (movekevin) [inline aptos-move/framework/src/natives/account.rs:27]: Can you leave a TODO here to move this into the gas schedule once we move it from upstream move into our repo?

File: aptos-move/framework/src/natives/account.rs [hunk 9]
```diff
@@ -16,29 +16,36 @@ use move_deps::{
...
-    arguments: VecDeque<Value>,
+    mut arguments: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
     debug_assert!(arguments.len() == 1);
 
-    let cost = native_gas(context.cost_table(), NativeCostIndex::DESTROY_SIGNER, 0);
-    Ok(NativeResult::ok(cost, smallvec![]))
+    let address = pop_arg!(arguments, AccountAddress);
+    let cost = native_gas(context.cost_table(), NativeCostIndex::CREATE_SIGNER, 0);
```
Comment (movekevin) [inline aptos-move/framework/src/natives/account.rs:49]: It's existing code but why is the cost 0 here?

File: config/management/genesis/src/storage_helper.rs [hunk 10]
```diff
@@ -68,11 +65,23 @@ impl StorageHelper {
         storage
             .import_private_key(FULLNODE_NETWORK_KEY, Ed25519PrivateKey::generate(&mut rng))
             .unwrap();
+        let owner_key = Ed25519PrivateKey::generate(&mut rng);
```
Comment (movekevin) [inline config/management/genesis/src/storage_helper.rs:68]: Are the changes in this file related?

File: config/management/genesis/src/command.rs [hunk 11]
```diff
@@ -325,74 +181,6 @@ pub mod tests {
         assert_eq!(layout_text, stored_layout);
     }
 
-    #[test]
-    fn test_validator_config() {
```
Comment (movekevin) [inline config/management/genesis/src/command.rs:329]: Why delete this?

File: config/management/genesis/src/command.rs [hunk 12]
```diff
@@ -139,10 +133,6 @@ impl Command {
         execute_command!(self, Command::SetOperator, CommandName::SetOperator)
     }
 
-    pub fn validator_config(self) -> Result<Transaction, Error> {
```
Comment (movekevin) [inline config/management/genesis/src/command.rs:142]: Why delete this?

File: aptos-move/vm-genesis/src/lib.rs [hunk 13]
```diff
@@ -378,27 +371,22 @@ impl TestValidator {
     pub fn new_test_set(count: Option<usize>) -> Vec<TestValidator> {
         let mut rng: rand::rngs::StdRng = rand::SeedableRng::from_seed([1u8; 32]);
         (0..count.unwrap_or(10))
-            .map(|idx| TestValidator::gen(idx, &mut rng))
+            .map(|_| TestValidator::gen(&mut rng))
             .collect()
     }
 
-    fn gen(index: usize, rng: &mut rand::rngs::StdRng) -> TestValidator {
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:385]: Why delete this?

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 6 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Account.move:354]: Per offline discussion, this can be left until later. @davidiw can you create an issue to track this so we don't forget?

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 3 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Account.move:380]: Same as previous comment

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/Account.move [hunk 14]
```diff
@@ -343,18 +343,95 @@ module AptosFramework::Account {
...
+    /// ```
+    /// public fun add_coin<X, Y>(lp: &LP, x: Coin<x>, y: Coin<y>) {
+    ///     if(!exists<LiquidityCoin<X, Y>(LP::Address(lp), LiquidityCoin<X, Y>)) {
+    ///         let mint, burn = Coin::initialize<LiquidityCoin<X, Y>>(...);
+    ///         move_to(&create_signer_with_capability(&lp.cap), LiquidityCoin<X, Y>{ mint, burn });
+    ///     }
+    ///     ...
+    /// }
+    /// ```
+    public fun create_resource_account(
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/Account.move:379]: How to prevent one of the signer from transferring all resource out of this account since signer has the highest access to all resources under the account? 
I think you mention it would be a separate PR. But what would the solution look like?

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:379]: In this case, the module / resource that owns this signer defines where the signer can be used. So the security is up to the author of those contracts. Much like the security of a signer is up to the owner of that account. We can add some awareness checks, but ultimately it is a user facing problem.

Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/Account.move:379]: Our current coin and token contracts allow the signer to transfer the resource. Does it mean we need to modify existing contract functions so that these functions verify the capability instead of signer?

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 5 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:38]: Hmm... Lemme follow up on this. It is late and I want to make the cut.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 6 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:354]: @areshand, this is something we should discuss tomorrow.

File: aptos-move/framework/aptos-framework/sources/Account.move (hunk 14 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Account.move:379]: nah, this is more for a shared account concept via a dao or something other module / resource sharing.

**Review by davidiw — COMMENTED** [review]
/land
